### ISSUE TITLE
[Bug] Swimlane profiling drops fanout edges for producers completing before consumer wiring

### ISSUE THREAD

**Issue by Hzfengsy** [ChaoWao/simpler#599]
## Platform

All (a2a3, a5)

## Runtime Variant

tensormap_and_ringbuffer (also affects aicpu_build_graph)

## Description

When AICPU profiling is enabled, the swimlane trace's per-task `fanout` list
silently drops dependency edges for any producer that has already transitioned
to `PTO2_TASK_COMPLETED` by the time the scheduler wires its downstream
consumer. The dropped edges never appear in `perf_swimlane_*.json` (and
therefore not in the Perfetto-converted trace either), which makes the
captured dependency graph a lower bound — it is structurally incomplete.

This was discovered while inspecting a Qwen3-32B single-layer decode run from
PyPTO. The decoder's scope-2 QK pre-matmul task (`func_12`) has 64 structural
producers from scope-2's K/V/Q RoPE stage (`func_11`). Only 16 of the 64
appear in the swimlane JSON's `fanout` lists — the other 48 have
`"fanout": []` and `"fanout_count": 0`:

```
func_11 tasks (distinct) : 64
  ├─ 16 with non-empty fanout (list r2t3 = the func_12 consumer)
  └─ 48 with empty fanout (none of the 64 → func_12 edges captured here)
```

Consumers of the dropped producers see incorrect (under-reported) fanout in
the trace UI, and any downstream tooling that reasons about the graph
(critical path, edge counts, flow arrows) is wrong by construction.

## Root Cause

The scheduler's consumer-wiring path in
`src/{a2a3,a5}/runtime/tensormap_and_ringbuffer/runtime/pto_scheduler.h:555-566`
only appends a consumer to the producer's `fanout_head` linked list when the
producer is **not yet** `>= PTO2_TASK_COMPLETED`. When the producer is
already completed, the wiring takes the `early_finished++` fast path and
**never** records the edge anywhere:

```cpp
pto2_for_each_fanin_slot_state(*wp, [&](PTO2TaskSlotState *producer) {
    pto2_fanout_lock(*producer);
    int32_t pstate = producer->task_state.load(std::memory_order_acquire);
    if (pstate >= PTO2_TASK_COMPLETED) {
        early_finished++;                              // ← edge not recorded
    } else {
        producer->fanout_head = rss.dep_pool.prepend(producer->fanout_head, ws);
    }
    pto2_fanout_unlock(*producer);
});
```

The AICPU profiler at
`src/{a2a3,a5}/runtime/tensormap_and_ringbuffer/aicpu/aicpu_executor.cpp:539-545`
walks `slot_state.fanout_head` and emits its entries into both the `fanout`
array and its `fanout_count`:

```cpp
uint64_t fanout_arr[RUNTIME_MAX_FANOUT];
int32_t fanout_n = 0;
PTO2DepListEntry *cur = slot_state.fanout_head;        // ← subset only
while (cur != nullptr && fanout_n < RUNTIME_MAX_FANOUT) {
    fanout_arr[fanout_n++] = cur->slot_state->task->task_id.raw;
    cur = cur->next;
}
perf_aicpu_complete_record(..., fanout_arr, fanout_n);
```

Note that `slot_state.fanout_count` (the *true* consumer count maintained by
the orchestrator at
`src/.../runtime/pto_orchestrator.cpp:731-735`) is correct — but the profiler
does not use it, and the output JSON's `fanout_count` field is the walk
length, not that true count.

`aicpu_build_graph` has the same structure at
`src/a2a3/runtime/aicpu_build_graph/runtime/pto_orchestrator.cpp:497-499`.

## Steps to Reproduce

1. Check out current `main` (commit `1d4c218`).
2. Use the attached `repro_fanout_lost.py` (see Additional Context). It
   builds a small pypto program with 64 parallel producers (`func_id=0`)
   fanning into one consumer (`func_id=1`).
3. Build & run with profiling:
   ```bash
   export PYTHONPATH=$PWD/python:$PYTHONPATH
   python repro_fanout_lost.py --runtime-profiling -p a2a3
   ```
4. Inspect the latest `build_output/ReproFanoutLost_*/swimlane_data/perf_swimlane_*.json`:
   ```bash
   python repro_fanout_lost.py --analyze build_output/ReproFanoutLost_*/swimlane_data/perf_swimlane_*.json
   ```

Alternatively, the issue can be observed on the Qwen3 decode example in the
PyPTO repo (`qwen3_bak.py`) by running with `--runtime-profiling` and looking
at `func_11` producers whose `"fanout": []`.

## Expected Behavior

Each of the 64 `func_id=0` producers in the repro lists the single
`func_id=1` consumer task id in its `fanout` array, and `fanout_count == 1`.

Equivalently for Qwen3: all 64 `func_11` producers of `r2t3` list that task
in their fanout.

## Actual Behavior

Roughly half (timing-dependent) of the `func_id=0` producers in the repro —
and 48 of the 64 `func_11` producers in Qwen3 — have `"fanout": []` and
`"fanout_count": 0`, even though they are structural producers.

Excerpt from the analyzer on a real Qwen3 trace:

```
func_id=11: 64 tasks, 16 with non-empty fanout, 48 with empty fanout
```

## Suggested Fix

Record the fanout edge on the **consumer-wiring path**, not on the
producer-completion path. Two low-cost shapes:

1. In `pto_scheduler.h` (both runtimes), extend the `early_finished++`
   branch to also append the consumer to a parallel, profiling-only list on
   the producer — e.g. reuse a separate `fanout_profile_head` list that is
   unconditionally populated regardless of producer state, and walked by the
   profiler instead of `fanout_head`.

2. Alternatively, populate the fanout list eagerly in
   `pto_orchestrator.cpp` at the same site where `producer->fanout_count`
   is incremented (lines 731-735). This is the natural "fanout_count and
   fanout_ids stay in sync" invariant and removes the race entirely, at the
   cost of moving the `dep_pool` allocation from the scheduler's wiring
   queue back onto the orchestrator's hot path (mitigated by using a
   profiling-only pool).

Option (1) is preferable because it keeps the orchestrator hot path free of
lock acquisition and leaves the existing `fanout_head` readiness-notification
mechanism untouched. Whichever path is picked, the profiler should walk the
new list (not `fanout_head`) and report `slot_state.fanout_count` directly
for the `fanout_count` JSON field.

## Git Commit ID

1d4c218456451fa99de3ace9601b01fd210a31dd

## Host Platform

Linux (aarch64)

## Additional Context

Minimal reproducer script (save as `repro_fanout_lost.py` in a pypto
checkout and run as described above):

```python
"""Reproducer: runtime swimlane trace drops fanout edges when a producer
completes before its consumer is wired (the `early_finished` fast path in the
scheduler).
"""
import argparse
import json
import sys
from pathlib import Path

N_PRODUCERS = 64
ROW = 64


def build_program():
    import pypto.language as pl

    @pl.program
    class ReproFanoutLost:
        @pl.function(type=pl.FunctionType.Opaque)
        def repro(
            self,
            a: pl.Tensor[[N_PRODUCERS, ROW], pl.FP32],
            out: pl.Out[pl.Tensor[[1, ROW], pl.FP32]],
        ) -> pl.Tensor[[1, ROW], pl.FP32]:
            tmp = pl.create_tensor([N_PRODUCERS, ROW], dtype=pl.FP32)
            # Scope 1: N parallel fast producers (func id 0).
            for i in pl.parallel(N_PRODUCERS):
                with pl.at(level=pl.Level.CORE_GROUP):
                    row = pl.mul(pl.slice(a, [1, ROW], [i, 0]), 2.0)
                tmp = pl.assemble(tmp, row, [i, 0])

            # Scope 2: single consumer (func id 1) fans in from all N producers.
            with pl.at(level=pl.Level.CORE_GROUP):
                acc = pl.full([1, ROW], dtype=pl.FP32, value=0.0)
                for i in pl.range(N_PRODUCERS):
                    acc = pl.add(acc, pl.slice(tmp, [1, ROW], [i, 0]))
            out = pl.assemble(out, acc, [0, 0])
            return out

    return ReproFanoutLost


def run(platform, device, runtime_profiling):
    import torch
    from pypto import ir
    from pypto.backend import BackendType
    from pypto.runtime import RunConfig

    backend = BackendType.Ascend950 if platform.startswith("a5") else BackendType.Ascend910B
    compiled = ir.compile(build_program(), backend_type=backend, platform=platform)
    a = torch.randn(N_PRODUCERS, ROW, dtype=torch.float32)
    out = torch.zeros(1, ROW, dtype=torch.float32)
    compiled(a, out, config=RunConfig(platform=platform, device_id=device,
             backend_type=backend, runtime_profiling=runtime_profiling))


def analyze(json_path):
    data = json.load(open(json_path))
    tasks = data["tasks"]
    by_func = {}
    for t in tasks:
        by_func.setdefault(t["func_id"], []).append(t)
    for fid in sorted(by_func):
        g = by_func[fid]
        empty = sum(1 for t in g if not t.get("fanout"))
        print(f"func_id={fid}: {len(g)} tasks, {len(g) - empty} non-empty fanout, {empty} empty")
    producers, consumers = by_func.get(0, []), by_func.get(1, [])
    if len(consumers) == 1:
        cid = consumers[0]["task_id"]
        feeders = [p for p in producers if cid in (p.get("fanout") or [])]
        print(f"Producers listing consumer in fanout: {len(feeders)}/{len(producers)}")


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("-p", "--platform", default="a2a3")
    ap.add_argument("-d", "--device", type=int, default=0)
    ap.add_argument("--runtime-profiling", action="store_true")
    ap.add_argument("--analyze", type=Path)
    args = ap.parse_args()
    if args.analyze:
        analyze(args.analyze)
    else:
        run(args.platform, args.device, args.runtime_profiling)
```

### Empirical data points (Qwen3-32B decode, a2a3)

```
Distinct func_11 tasks : 64
  non-empty fanout     : 16  (tasks r3t866, r3t869, ... r3t911, stride 3)
  empty fanout (= [])  : 48  (tasks r3t722 ... r3t863)
```

All 64 are structural producers of `func_12_m(r2t3)` in the IR, but only
the 16 that were still running when `r2t3` was wired appear in its
inverse-fanout.

### Impact

The swimlane trace is the primary tool used when debugging compile / runtime
regressions involving parallelism and cross-scope dependencies. Silently
dropping edges causes misleading timelines (missing flow arrows), undercounts
in sched-overhead analysis, and wrong critical-path reasoning when dependency
structure is inferred from the trace.

**Comment by ChaoWao** [comment]
Good points. The runtime build graph and execute togather. So it can hardly profiling full information when running. We will profiling those information using a simulation run method to fully build the graph. After that, combining with the profiling execution time to generate no-info-lost swimlane.
